File: Power Query/300-399/389/PQ_389.R
library(dplyr)
library(readxl)
library(slider)

path <- "300-399/389/PQ_Challenge_389.xlsx"
input <- read_excel(path, range = "A1:C29")
test <- read_excel(path, range = "E1:H25")

result <- input |>
  transmute(
    ID,
    Pattern = slide_chr(
      Code,
      paste,
      collapse = "",
      .before = 2,
      .after = 2,
      .complete = TRUE
    ),
    WindowsValueSum = slide_dbl(
      Value,
      sum,
      .before = 2,
      .after = 2,
      .complete = TRUE
    ),
    DominantCode = slide_chr(
      Code,
      \(x) {
        t <- table(x)
        names(t)[which.max(t)]
      },
      .before = 2,
      .after = 2,
      .complete = TRUE
    )
  ) |>
  na.omit()

all.equal(result, test, check.attributes = FALSE)
# TRUE

File: Power Query/300-399/389/PQ_389.py
import pandas as pd
from collections import Counter

path = "300-399/389/PQ_Challenge_389.xlsx"
input = pd.read_excel(path, usecols="A:C", nrows=29)
test = pd.read_excel(path, usecols="E:H", nrows=24).rename(
	columns=lambda c: c.replace(".1", "") if isinstance(c, str) else c
)

for df in (input, test):
	n = df.select_dtypes(include="number").columns
	df[n] = df[n].astype(float)
result = input.copy()
windows = [slice(i - 2, i + 3) if 1 < i < len(result) - 2 else None for i in range(len(result))]
result["sum"] = [result["Value"].iloc[w].sum() if w else pd.NA for w in windows]
result["conc"] = ["".join(result["Code"].iloc[w].astype(str)) if w else pd.NA for w in windows]
result["most_common"] = [
	sorted(Counter(result["Code"].iloc[w]).items(), key=lambda x: (-x[1], str(x[0])))[0][0] if w else pd.NA
	for w in windows
]
result = result.dropna().loc[:, ["ID", "conc", "sum", "most_common"]].rename(
	columns={"conc": "Pattern", "sum": "WindowValueSum", "most_common": "DominantCode"}
).reset_index(drop=True)

print(all(result == test))
